cpp/src/stream_compaction/copy_if.cuh

```diff
@@ -326,51 +343,68 @@ gdf_column copy_if(gdf_column const &input, Filter filter,
   }
 
   CHECK_STREAM(stream);
-
-  gdf_column output = cudf::empty_like(input);
   
   // 3. compute the output size from the last block's offset + count
   gdf_size_type output_size = 
     get_output_size(block_counts, block_offsets, grid.num_blocks, stream);
 
-  if (output_size > 0) {    
-    // Allocate/initialize output column
-    auto column_byte_width { cudf::byte_width(input) };
-
-    void *data = nullptr;
-    gdf_valid_type *valid = nullptr;
-    RMM_ALLOC(&data, output_size * column_byte_width, stream);
+  table output;
 
-    if (input.valid != nullptr) {
-      auto bytes = gdf_valid_allocation_size(output_size);
-      RMM_ALLOC(&valid, bytes, stream);
-      CUDA_TRY(cudaMemsetAsync(valid, 0, bytes, stream));
-    }
-
-    CUDF_EXPECTS(GDF_SUCCESS == gdf_column_view(&output, data, valid,
-                                                output_size, input.dtype),
-                "cudf::apply_boolean_mask failed to create output column view");
+  if (output_size == input.num_rows()) {
+    output = cudf::copy(input);
+  }
+  else if (output_size > 0) {
+    // Allocate/initialize output columns
+    output = cudf::allocate_like(input, output_size, true, stream);
 
     // 4. Scatter the output data and valid mask
-    cudf::type_dispatcher(output.dtype,
-                          scatter_functor<Filter, block_size, per_thread>{},
-                          output, input, block_offsets, filter,
-                          input.valid != nullptr, stream);
-
-    CHECK_STREAM(stream);
+    for (int col = 0; col < input.num_columns(); col++) {
+      cudf::type_dispatcher(output.get_column(col)->dtype,
+                            scatter_functor<Filter, block_size, per_thread>{},
+                            *output.get_column(col), *input.get_column(col), 
+                            block_offsets, filter, stream);
+    }
   }
-  
-  // synchronize nvcategory after filtering
-  if (output.dtype == GDF_STRING_CATEGORY) {
-    CUDF_EXPECTS(
-    GDF_SUCCESS ==
-      nvcategory_gather(&output,
-                        static_cast<NVCategory *>(input.dtype_info.category)),
-      "could not set nvcategory");
+  else {
+    output = empty_like(input);
   }
+  
+  CHECK_STREAM(stream);
+
+  gdf_error err = nvcategory_gather_table(input, output);
+  CHECK_STREAM(0);
+  CUDF_EXPECTS(err == GDF_SUCCESS, "Error updating nvcategory in copy_if");
   return output;
 }
 
+/*
+ * @brief Filters a column using a Filter function object
+ * 
+ * @p filter must be a functor or lambda with the following signature:
+ * __device__ bool operator()(gdf_index_type i);
+ * It returns true if element i of @p input should be copied, false otherwise.
+ * 
+ * @note: filter must return false if its input i is out of bounds for any 
```

Comment:
What's the reasoning behind this? Is it difficult to put the bounds check into the kernel?

As a hypothetical user of this function, I'd find this surprising as I'd expect it to work like `std::copy_if` or `thrust::copy_if` where the predicate does not need to perform it's own bounds checking.

Reply:
Done.